upload/admin/model/extension/smailyforopencart/admin.php

```diff
@@ -30,73 +30,218 @@ public function uninstall() {
+            $this->log->write($error->getMessage());
+        }
+
+        if (empty($autoresponders)) {
+            return array();
+        }
+
+        $list = [];
+        foreach ($autoresponders as $autoresponder) {
+            if (!empty($autoresponder['id']) && !empty($autoresponder['title'])) {
+                $list[$autoresponder['id']] = trim($autoresponder['title']);
             }
-            $api_call = json_decode(curl_exec($ch), true);
-            // Response code from Smaily API.
-            $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
-        // If Method POST.
-        } elseif ($method === 'POST') {
-            curl_setopt($ch, CURLOPT_POST, 1);
-            curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($data));
-            $api_call = json_decode(curl_exec($ch), true);
-            // Response code from Smaily API.
-            $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
-            // Validate response
-            if (!array_key_exists('code', $api_call)) {
-                $this->log->write('Something went wrong with Smaily request!');
+        }
+        return $list;
+    }
+
+    /**
+     * Helper function to validate credentials provided in argument.
+     *
+     * @param string $subdomain Smaily API subdomain.
+     * @param string $username Smaily API username.
+     * @param string $password Smaily API password.
+     * @return array $response Status message.
+     */
+    public function validateCredentials($subdomain, $username, $password) {
+        $this->load->language('extension/module/smaily_for_opencart');
+        // In case subdomain was entered as http://demo.sendsmaily.net.
+        $subdomain = $this->normalizeSubdomain($subdomain);
+        $username = html_entity_decode($username);
+        $password = html_entity_decode($password);
+
+        $response = array();
+        // Validate credentials with a call to Smaily.
+        try {
+            $response = (new \Smaily\Request)
+                ->auth($subdomain, $username, $password)
+                ->setUrlViaEndpoint('workflows')
+                ->setData(array('trigger_type' => 'form_submitted'))
+                ->get();
+            $this->saveValidatedCredentials($subdomain, $username, $password);
+            $response['success'] = $this->language->get('validated_success');
+        } catch(Smaily\HTTPError $error) {
+            switch($error->getCode()) {
+                case Smaily\Request::HTTP_ERR_UNAUTHORIZED:
+                    $response['error'] = $this->language->get('validated_unauthorized');
+                    break;
+
+                case Smaily\Request::HTTP_ERR_INVALID_SUBDOMAIN:
+                    $response['error'] = $this->language->get('validated_subdomain_error');
+                    break;
+
+                default:
+                    $response['error'] = $this->language->get('validated_error');
+            }
+        }
+        return $response;
```

Comment:
Let's clean up the `$response` variable usage a bit. Right now it is a mash-up of both API and actual frontend response. It could accidentally leak sensitive information to the frontend.

```suggestion
        try {
            (new \Smaily\Request)
                ->auth($subdomain, $username, $password)
                ->setUrlViaEndpoint('workflows')
                ->setData(array('trigger_type' => 'form_submitted'))
                ->get();
        } catch(Smaily\HTTPError $error) {
            $response = array();
        
            switch($error->getCode()) {
                case Smaily\Request::HTTP_ERR_UNAUTHORIZED:
                    $response['error'] = $this->language->get('validated_unauthorized');
                    break;

                case Smaily\Request::HTTP_ERR_INVALID_SUBDOMAIN:
                    $response['error'] = $this->language->get('validated_subdomain_error');
                    break;

                default:
                    $response['error'] = $this->language->get('validated_error');
            }
            
            return $response;
        }
        
        $this->saveValidatedCredentials($subdomain, $username, $password);
        
        return ['success' => $this->language->get('validated_success')];
```

`saveValidatedCredentials` won't throw an `APIError` or a `HTTPError` anyway so no point in keeping it inside the try-catch statement.

Reply:
> saveValidatedCredentials won't throw an APIError or a HTTPError anyway so no point in keeping it inside the try-catch statement.

But I only want to save credentials when the Request doesn't throw an error. I kept it in the try block to make sure only valid credentials are saved

Reply:
I think I removed setting the new Request to a variable in earlier commits already. Added commit 6a6449c to simplify the code. I think I simplified the quoted lines as much as possible, but if you see more ways or something I missed, please point them out to me.

BTW this code was moved to controller and should be `ajaxValidateCredentials()`

Reply:
When you bring `saveValidatedCredentials` outside of the exception it will still save credentials only when `Request` callback succeeds, because any other exception that try-catch does not capture (i.e. is not `Smaily\HTTPError`) will still be thrown. With `catch` statement you simply tell PHP how to handle these type of errors.